### PR TITLE
Mapping

### PR DESCRIPTION
**Pull request by justinlittman**
## Why was this change made?
To map from Orcid works to pub hashes and vice versa.

## How was this change tested?
Unit

## Which documentation and/or configurations were updated?
NA

### CODE REVIEW

File: lib/csl/citation.rb [hunk 1]
```diff
@@ -1,26 +1,12 @@
 # frozen_string_literal: true
 
 require 'citeproc/ruby'
-# You can install more recent styles using the 'csl-styles' gem or you can just
-# download the styles you need (just point CSL::Style.root to the directory and
-# you can load any style with just its name using CSL::Style.load).
-require 'csl/styles'
```
Comment (justinlittman) [inline lib/csl/citation.rb:7]: Moving rendering into its own class so that can be re-used.

File: lib/csl/mapper.rb [hunk 2]
```diff
@@ -59,7 +59,7 @@ def csl_doc
         # Add journal information if it exists.
         if pub_hash.key?(:journal)
           if pub_hash[:journal][:name].present?
-            cit_data_hash['type'] = 'article-journal'
+            cit_data_hash['type'] = 'article-journal' unless cit_data_hash['type'] == 'inproceedings'
```
Comment (justinlittman) [inline lib/csl/mapper.rb:62]: Arcadia noticed that we were incorrect setting type for `inproceedings`.

File: lib/orcid/add_works.rb [hunk 3]
```diff
@@ -41,6 +41,9 @@ def add_work(author, contribution, orcid_user)
       contribution.save!
       logger&.info("#{self.class} - author #{author.id} - added publication #{contribution.publication.id} with put-code #{contribution.orcid_put_code}")
       true
+    rescue Orcid::PubMapper::PubMapperError => e
```
Comment (justinlittman) [inline lib/orcid/add_works.rb:44]: PubMapperErrors are worthy of logging, but not notifying.

File: lib/orcid/pub_author_mapper.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        }
+      }
+    end
+
+    private
+
+    attr_reader :author_hash
+
+    def map_credit_name
+      clean_name(author_hash[:name].presence) ||
```
Comment (justinlittman) [inline lib/orcid/pub_author_mapper.rb:36]: We take various strategies to constructing a name. The ordering of those strategies is deliberate.

File: lib/orcid/pub_identifier_mapper.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    def map
+      ids = map_identifiers(pub_hash[:identifier], 'self') +
+            map_identifiers(pub_hash.dig(:journal, :identifier), 'part-of') +
+            map_identifiers(pub_hash.dig(:conference, :identifier), 'part-of') +
+            map_identifiers(pub_hash.dig(:series, :identifier), 'part-of')
+
+      raise PubMapper::PubMapperError, 'An identifier is required' if ids.empty?
+
+      clean_issns(ids)
```
Comment (justinlittman) [inline lib/orcid/pub_identifier_mapper.rb:25]: A couple identifier pecularities cause us problems, so we clean them here.

File: lib/csl/citation.rb [hunk 6]
```diff
@@ -31,43 +17,28 @@ def initialize(pub_hash)
     # @return [Hash]
     def citations
       {
-        apa_citation: to_apa_citation,
-        mla_citation: to_mla_citation,
-        chicago_citation: to_chicago_citation
+        apa_citation: renderer.to_apa_citation,
+        mla_citation: renderer.to_mla_citation,
+        chicago_citation: renderer.to_chicago_citation
```
Comment (peetucket) [inline lib/csl/citation.rb:22]: based on the delegation above in line 9, is calling `renderer.` here still necessary?

File: lib/csl/mapper.rb (hunk 2 above)
Comment (peetucket) [reply lib/csl/mapper.rb:62]: this is used when mapping pub_hash to the citation generator right?  so this basically would change for how we format citations going forward

File: lib/orcid/add_works.rb (hunk 3 above)
Comment (peetucket) [reply lib/orcid/add_works.rb:44]: so is the idea that we monitor the logs for a bit to see the types of unexpected mapping issues that are occurring?

**Review by peetucket — COMMENTED** [review]
Initial pass with some comments...

File: lib/csl/mapper.rb (hunk 2 above)
Comment (justinlittman) [reply lib/csl/mapper.rb:62]: I suspect it doesn't actually change any MLA, Chicago, or APA citations, but will only affect Bibtex citations (which are new).

File: lib/orcid/add_works.rb (hunk 3 above)
Comment (justinlittman) [reply lib/orcid/add_works.rb:44]: We should keep in eye out, but this is primarily going to report issues like "this publication doesn't have an identifier." Any significant issues will be HB alerted.

File: lib/csl/citation.rb (hunk 6 above)
Comment (justinlittman) [reply lib/csl/citation.rb]: Nope. Removed.

File: spec/lib/orcid/pub_identifier_mapper_spec.rb [hunk 7]
```diff
@@ -246,6 +258,12 @@
 
       it 'ignores self identifier' do
```
Comment (jmartin-sul) [inline spec/lib/orcid/pub_identifier_mapper_spec.rb:259]: is this test title still accurate?

**Review by jmartin-sul — CHANGES_REQUESTED** [review]
only issues i noticed were a couple test titles that might need fixing, otherwise 👍

File: spec/lib/orcid/pub_identifier_mapper_spec.rb (hunk 7 above)
Comment (justinlittman) [reply spec/lib/orcid/pub_identifier_mapper_spec.rb]: Accurate, but not fully descriptive. Fixed.

File: spec/lib/orcid/add_works_spec.rb [hunk 8]
```diff
@@ -68,12 +68,7 @@
       # Default factory pub_hash is missing an identifier so allowing factory to create publication.
       let!(:contribution) { create :contribution, author: author } # rubocop:disable RSpec/LetSetup
 
-      before do
-        allow(NotificationManager).to receive(:error)
-      end
-
       it 'notifies' do
```
[the remark this answers is not in the document]
Comment (justinlittman) [reply spec/lib/orcid/add_works_spec.rb]: Fixed and added `NotificationManager` check (since NotificationManager is mocked, I don't think it is overkill).

**Review by jmartin-sul — APPROVED** [review]

**Comment by jmartin-sul** [comment]
@justinlittman / @peetucket -- wasn't sure if peter wanted to do more review, so i held off on merging, but approved, feel free to merge if sufficiently reviewed.

**Comment by peetucket** [comment]
I'm ok without a re-review if you are both satisfied.

**Comment by jmartin-sul** [comment]
> I'm ok without a re-review if you are both satisfied.

merged!

**Pull request merged**